server/src/main/java/org/elasticsearch/index/analysis/AnalysisRegistry.java

```diff
@@ -544,4 +530,80 @@ private void processNormalizerFactory(
+        NamedAnalyzer newDefaultSearchQuoteAnalyzer = rebuildIfNecessary(indexAnalyzers.getDefaultSearchQuoteAnalyzer(), indexSettings);
+        Map<String, NamedAnalyzer> newAnalyzers = new HashMap<>();
+        for (NamedAnalyzer analyzer : indexAnalyzers.getAnalyzers().values()) {
+            newAnalyzers.put(analyzer.name(), rebuildIfNecessary(analyzer, indexSettings));
+        }
+        return new IndexAnalyzers(indexSettings, indexAnalyzers.getDefaultIndexAnalyzer(), newDefaultSearchAnalyzer,
+                newDefaultSearchQuoteAnalyzer, newAnalyzers, indexAnalyzers.getNormalizers(), indexAnalyzers.getWhitespaceNormalizers());
+    }
+
+    /**
+     * Check if the input analyzer needs to be rebuilt. If not, return analyzer unaltered, otherwise rebuild it. We currently only consider
+     * instances of {@link CustomAnalyzer} with {@link AnalysisMode#SEARCH_TIME} to be eligible for rebuilding.
+     */
+    private NamedAnalyzer rebuildIfNecessary(NamedAnalyzer oldAnalyzer, IndexSettings indexSettings)
+            throws IOException {
+        // only rebuild custom analyzers that are in SEARCH_TIME mode
+        if ((oldAnalyzer.getAnalysisMode() == AnalysisMode.SEARCH_TIME) == false
+                || (oldAnalyzer.analyzer() instanceof CustomAnalyzer == false)) {
+            return oldAnalyzer;
+        } else {
+            String analyzerName = oldAnalyzer.name();
+
+            // get tokenizer necessary to re-build the analyzer
+            String tokenizer = indexSettings.getSettings().get("index.analysis.analyzer." + analyzerName + ".tokenizer");
+            Map<String, AnalysisProvider<TokenizerFactory>> tokenizerProvider = Collections.singletonMap(tokenizer,
+                    getTokenizerProvider(tokenizer, indexSettings));
+            final Map<String, Settings> tokenizerSettings = indexSettings.getSettings().getGroups(INDEX_ANALYSIS_TOKENIZER);
+            final Map<String, TokenizerFactory> tokenizers = buildMapping(Component.TOKENIZER, indexSettings, tokenizerSettings,
+                    tokenizerProvider, Collections.emptyMap());
+
+            // get char filters necessary to re-build the analyzer
+            List<String> charFilterNames = indexSettings.getSettings()
+                    .getAsList("index.analysis.analyzer." + analyzerName + ".char_filter");
+            Map<String, AnalysisProvider<CharFilterFactory>> charFilterProvider = charFilterNames.stream()
+                    .map(s -> new AbstractMap.SimpleEntry<>(s, getCharFilterProvider(s, indexSettings)))
+                    .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+            final Map<String, Settings> charFiltersSettings = indexSettings.getSettings().getGroups(INDEX_ANALYSIS_CHAR_FILTER);
+            final Map<String, CharFilterFactory> charFilters = buildMapping(Component.CHAR_FILTER, indexSettings,
+                    charFiltersSettings, charFilterProvider, Collections.emptyMap());
```

Comment:
I hope the above can still be reduced, I was e.g. toying with the ides of adding references to the  `Map<String, XYZFactory>` for tokenizers and charfilters to IndexAnalyzers to re-use them when rebuilding. we create them when initially creating IndexAnalyzers in AnalysisRegistry#build but then throw them away. Since they shouldn't change over time (at least the tokenizer and char filters, maybe also part of the token filters) we could reuse them if they were available via IndexAnalyzers. I didn't go down this road yet though.